src/main/java/com/google/sps/data/Chat.java

```diff
@@ -42,54 +42,126 @@
+
+        if (chat_exists == true){
+            return null;
+        }
+
+        KeyFactory keyFactory = datastore.newKeyFactory().setKind("chat");
+        Key chatKey = datastore.allocateId(keyFactory.newKey());
+
+        Entity chatEntity = Entity.newBuilder(chatKey)
+            .set("participants", participantId1, participantId2)
+            .set("messages", ListValue.of("Chat Start"))
+            .build();
 
-    try {
-      Entity entity = datastore.get(key);
-      Chat c = new Chat(entity);
-      return c;
-    } catch (Exception e) {
-      return null;
+        datastore.add(chatEntity);
+        System.out.println("Breakpoint");
+        Chat chat = new Chat(chatEntity);
+
+        return chat;
     }
-  }
 
-  public static List<Chat> getChats(Datastore datastore, String participantid) {
-    
-    List<Chat> chats = new ArrayList<>();
-    Query<Entity> chatQuery = Query
-    .newEntityQueryBuilder()
-    .setKind("Chat")
-    .build();
-    QueryResults<Entity> chatEntities = datastore.run(chatQuery);
-
-    // Iterate through the chats
-    // for (Entity entity : chatEntities.asIterable()) {
-    //   long chatid = entity.getKey.getId();
-    //   List<String> participantList = Arrays.asList(entity.getProperty("participants").split(","));
-
-    //   if (participantList.contains(participantid)){
-    //     Chat chat = new Chat(entity);
-    //     chats.add(chat);
-    //   }
-    // }
-    return chats;
-  }
+    public static Chat getChatById(Datastore datastore, String chatId) {
+        KeyFactory keyFactory = datastore.newKeyFactory().setKind("chat");
 
+        Query<Entity> query = Query.newEntityQueryBuilder()
+            .setKind("chat")
+            .setFilter(PropertyFilter.eq("__key__", keyFactory.newKey(chatId)))
+            .build();
 
-  long getId(){
-    return id;
-  }
+        QueryResults<Entity> chatResults = datastore.run(query);
+        
+        try {
+            Entity chat = chatResults.next();
+            Chat c = new Chat(chat);
+            System.out.println("Chat found");
+            return c;
+        } catch (Exception e) {
+            System.out.println("Chat not found");
+            return null;
+        }
+    }
 
-  String getParticipants(){
-    return participants;
-  }
+    public static List<Chat> getChatsByUser(Datastore datastore, String participantId) {
+        
+        List<Chat> chats = new ArrayList<Chat>();
+        Query<Entity> chatQuery = Query.newEntityQueryBuilder().setKind("chat")
+        .setFilter(CompositeFilter.and(
+            PropertyFilter.eq("participants", participantId)))
+        .build();
+        QueryResults<Entity> chatEntities = datastore.run(chatQuery);
 
-  List<String> getMessages(){
-    return messages;
-  }
-}
\ No newline at end of file
+        while(chatEntities.hasNext()){
+            Entity entity = chatEntities.next();
+            Chat chat = new Chat(entity);
+            chats.add(chat);
+        }
```

Comment:
FYI, you could write this using the Iterator interface's [`forEachRemaining`](https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html#forEachRemaining-java.util.function.Consumer-):
```java
chatEntities.forEachRemaining(entity -> chats.add(new Chat(entity)));
```

I noticed you have this pattern in a couple other places too.

Reply:
Good note